pkg/ingester/checkpoint.go

```diff
@@ -59,3 +76,420 @@ func fromWireChunks(conf *Config, wireChunks []Chunk) ([]*chunkDesc, error) {
 	}
 	return descs, nil
 }
+
+// nolint:interfacer
+func decodeCheckpointRecord(rec []byte, s *Series) error {
+	//TODO(owen-d): reduce allocs
+	// The proto unmarshaling code will retain references to the underlying []byte it's passed
+	// in order to reduce allocs. This is harmful to us because when reading from a WAL, the []byte
+	// is only guaranteed to be valid between calls to Next().
+	// Therefore, we copy it to avoid this problem.
+	cpy := make([]byte, len(rec))
+	copy(cpy, rec)
+
+	switch RecordType(cpy[0]) {
+	case CheckpointRecord:
+		return proto.Unmarshal(cpy[1:], s)
+	default:
+		return errors.Errorf("unexpected record type: %d", rec[0])
+	}
+}
+
+func encodeWithTypeHeader(m proto.Message, typ RecordType, b []byte) ([]byte, error) {
+	buf, err := proto.Marshal(m)
+	if err != nil {
+		return b, err
+	}
+
+	b = append(b[:0], byte(typ))
+	b = append(b, buf...)
+	return b, nil
+}
+
+type SeriesWithErr struct {
+	Err    error
+	Series *Series
+}
+
+type SeriesIter interface {
+	Num() int
+	Iter() <-chan *SeriesWithErr
+	Stop()
+}
+
+type ingesterSeriesIter struct {
+	ing *Ingester
+
+	done chan struct{}
+}
+
+func newIngesterSeriesIter(ing *Ingester) *ingesterSeriesIter {
+	return &ingesterSeriesIter{
+		ing:  ing,
+		done: make(chan struct{}),
+	}
+}
+
+func (i *ingesterSeriesIter) Num() (ct int) {
+	for _, inst := range i.ing.getInstances() {
+		ct += inst.numStreams()
+	}
+	return ct
+}
+
+func (i *ingesterSeriesIter) Stop() {
+	close(i.done)
+}
+
+func (i *ingesterSeriesIter) Iter() <-chan *SeriesWithErr {
+	ch := make(chan *SeriesWithErr)
+	go func() {
+		for _, inst := range i.ing.getInstances() {
+			inst.streamsMtx.RLock()
+			// Need to buffer streams internally so the read lock isn't held trying to write to a blocked channel.
+			streams := make([]*stream, 0, len(inst.streams))
+			inst.streamsMtx.RUnlock()
+			_ = inst.forAllStreams(func(stream *stream) error {
+				streams = append(streams, stream)
+				return nil
+			})
+
+			for _, stream := range streams {
+				// TODO(owen-d): use a pool
+				chunks, err := toWireChunks(stream.chunks, nil)
```

Comment:
You most likely have a race here:
- 1 the list of streams can change
- 2 the list of chunks too.

You should write a test that run 2 goroutines, one that push entries to ingester and one that read this iterator,  then using `go test -race`. 

This is definitively tricky cause you want a snapshot of chunks but you're using a chanel which is driven by the reader and so in-between reads things can change.

I feel like you should either lock everything until iteration is over which seems dangerous or return an arrays of `SeriesWithErr`

Reply:
This is/will be covered by two things:
1) Streams are buffered to an intermediate channel during checkpoint start. 
This helps synchronize the perSeriesDuration with a list of chunks and avoids the changing problem. The caveat is an old stream may be kept around for up to the checkpoint interval (default 5m).
https://github.com/grafana/loki/blob/316618342cfccb64b8db22498e0228be339a4644/pkg/ingester/checkpoint.go#L149-L155
https://github.com/grafana/loki/blob/316618342cfccb64b8db22498e0228be339a4644/pkg/ingester/checkpoint.go#L438-L442

2) The locking is mainly covered in a followup PR I'm preparing.

Reply:
Please include a concurrent access tests in your following PR.